File: homework-1/frontend/src/app/models/transaction.ts
export type TransactionType = 'deposit' | 'withdrawal' | 'transfer';
export type TransactionStatus = 'pending' | 'completed' | 'failed';

export interface Transaction {
  id: string;
  fromAccount: string;
  toAccount: string;
  amount: number;
  currency: string;
  type: TransactionType;
  timestamp: string;
  status: TransactionStatus;
}

export interface TransactionRequest {
  fromAccount: string;
  toAccount: string;
  amount: number;
  currency: string;
  type: TransactionType;
}

export interface TransactionFilters {
  accountId?: string;
  type?: string;
  from?: string;
  to?: string;
}

export interface BalanceResponse {
  accountId: string;
  balance: number;
}

export interface SummaryResponse {
  totalDeposits: number;
  totalWithdrawals: number;
  transactionCount: number;
  mostRecentTransactionDate: string | null;
}

export interface InterestResponse {
  accountId: string;
  balance: number;
  rate: number;
  days: number;
  interest: number;
}

export interface ValidationDetail {
  field: string;
  message: string;
}

export interface ApiError {
  error: string;
  message?: string;
  details?: ValidationDetail[];
}

File: homework-1/backend-go/internal/model/transaction.go
package model

import (
	"time"

	"github.com/shopspring/decimal"
)

func init() {
	// Serialize monetary amounts as JSON numbers (e.g. 100.50), not quoted strings,
	// matching the API contract and the Angular client.
	decimal.MarshalJSONWithoutQuotes = true
}

type TransactionType string

const (
	Deposit    TransactionType = "deposit"
	Withdrawal TransactionType = "withdrawal"
	Transfer   TransactionType = "transfer"
)

type TransactionStatus string

const (
	Pending   TransactionStatus = "pending"
	Completed TransactionStatus = "completed"
	Failed    TransactionStatus = "failed"
)

// Transaction is the stored/returned entity. Money is a decimal (never float).
type Transaction struct {
	ID          string            `json:"id"`
	FromAccount string            `json:"fromAccount"`
	ToAccount   string            `json:"toAccount"`
	Amount      decimal.Decimal   `json:"amount"`
	Currency    string            `json:"currency"`
	Type        TransactionType   `json:"type"`
	Timestamp   time.Time         `json:"timestamp"`
	Status      TransactionStatus `json:"status"`
}

// TransactionRequest is the incoming payload. Type stays a string so validation can
// produce structured, per-field errors (added in Cycle 2) instead of failing to decode.
type TransactionRequest struct {
	FromAccount string          `json:"fromAccount"`
	ToAccount   string          `json:"toAccount"`
	Amount      decimal.Decimal `json:"amount"`
	Currency    string          `json:"currency"`
	Type        string          `json:"type"`
}
